Add integration test for display cutouts on Android screen rotation

yaakovschectman opened the pull request:
Test that the position of a cutout as reported by the Android engine repositions based on screen orientation.

Related to https://github.com/flutter/engine/pull/55992

Part of https://github.com/flutter/flutter/issues/155658

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

yaakovschectman committed:
Add integration test

yaakovschectman committed:
Format

yaakovschectman commented:
Until https://github.com/flutter/engine/pull/55992 lands, the newly added test *should* fail.

yaakovschectman committed:
Timeouts

yaakovschectman commented:
https://github.com/flutter/engine/pull/55992 landed, but the CI will fail until the engine changes roll into framework.

yaakovschectman commented:
Looks like something isn't working as nicely on CI as it is locally... I'll need to check that out

reidbaker commented:
I think you are hitting a timeout I am not sure of the default but assembly which appears to include downloading dependencies is taking  34 seconds `"Running Gradle task 'assembleDebug'...                             34.2s`

yaakovschectman committed:
Merge branch 'master' into android_it_rot

yaakovschectman commented:
Ah, probably should have rebased on master after the engine roll landed before worrying about the test failure. Let's see now.

yaakovschectman committed:
Set layout mode for API<35

yaakovschectman committed:
Merge branch 'master' into android_it_rot

yaakovschectman commented:
The problem *was* that, with API<35, the default behavior of an app is to make cutouts opaque to the application unless it requests otherwise, so we need to request that in order to access and test the cutouts.

Now, the problem *is* that `FlutterDriver.waitFor` never completes, even after its timeout, but only the first time the test is driven after starting an emulator (which is a problem for CI). *This* only happens when `adb` is called in `setupAll` to enable an overlay, which actually simulates the cutout that we then test.

If I could figure out a way to set up the display cutout *before* the test app launches, this would be easily resolved. But doing it in `setupAll` is a problem.

yaakovschectman committed:
Enable overlay before flutter drive

yaakovschectman committed:
Organize code

yaakovschectman committed:
Comment rationale for code

yaakovschectman committed:
Merge branch 'master' into android_it_rot

yaakovschectman commented:
I've moved the emulator setup out of `setupAll` into the dart script that runs the corresponding shard of tests. This way, `adb` enables the cutout before `flutter drive` is actually run, and the driver is not broken. I have also added comments to document the requirements of a device/emulator to run these new tests, as well as how and why those requirements are satisfied by our CI.

reidbaker committed:
Merge branch 'master' into android_it_rot

matanlurey commented on dev/bots/suite_runners/run_flutter_driver_android_tests.dart:
```diff
@@ -79,4 +79,49 @@ Future<void> runFlutterDriverAndroidTests() async {
     ),
   );
+
+  // A simulated cutout overlay is necessary to test display_cutout_rotation.
```
This has the possibility of polluting the test device, and/or making future tests added (or refactored) have confusing behavior. That being said, that's sort of true of _all_ `adb` commands, but at least it would be more obvious.

Let's make sure we toggle these settings off. I'd like to see a pattern like:

```dart
try {
  _enableDevelopmentSettings();
  _enableOverlay('com.android.internal.display.cutout.emulation.tall');
  runCommand('flutter', ['drive', ...]);
} finally {
  _disableDevelopmentSettings();
  _disableOverlay();
}
```

matanlurey commented on dev/integration_tests/native_driver_test/android/app/src/main/kotlin/com/example/native_driver_test/MainActivity.kt:
```diff
@@ -45,5 +47,15 @@ class MainActivity : FlutterActivity() {
         windowInsetsController.hide(WindowInsetsCompat.Type.systemBars())
         actionBar?.hide()
+
+        // The default behavior on SDK level 34 and below is for display cutouts to be consumed
```
But the version below is `>= 30`?

yaakovschectman replied:
`WindowManager.LayoutParams.LAYOUT_IN_DISPLAY_CUTOUT_MODE_ALWAYS` was introduced in 30, and as of 35 it is essentially the default behavior. So for SDKs 30<=x<=34 (and these run on a 34 emulator), this attribute needs to be set.

matanlurey commented on dev/integration_tests/native_driver_test/lib/display_cutout_rotation_main.dart:
```diff
@@ -0,0 +1,59 @@
+import 'src/allow_list_devices.dart';
+
+void main() async {
+  ensureAndroidOrIosDevice();
```
No longer accurate. Please add and use a `ensureAndroidDevice()` method.

matanlurey commented on dev/integration_tests/native_driver_test/lib/display_cutout_rotation_main.dart:
```diff
@@ -0,0 +1,59 @@
+
+class _MainAppState extends State<MainApp> {
+  @override
+  Widget build(BuildContext context) {
```
Please leave comments explaining what this is doing, it's really hard to understand where (or what) this does and exactly what we're checking for.

matanlurey commented on dev/integration_tests/native_driver_test/lib/display_cutout_rotation_main.dart:
```diff
@@ -0,0 +1,59 @@
+  Widget build(BuildContext context) {
+    final List<DisplayFeature> displayFeatures =
+        MediaQuery.of(context).displayFeatures;
+    displayFeatures.retainWhere(
```
Are we certain it's safe to mutate the result of `MediaQuery.of(...)`?

If not, I'd feel better with a `.where().toList()`.

matanlurey commented on dev/integration_tests/native_driver_test/lib/display_cutout_rotation_main.dart:
```diff
@@ -0,0 +1,59 @@
+        MediaQuery.of(context).displayFeatures;
+    displayFeatures.retainWhere(
+        (DisplayFeature feature) => feature.type == DisplayFeatureType.cutout);
+    String text;
```
nit: `final String`.

matanlurey commented on dev/integration_tests/native_driver_test/test_driver/display_cutout_rotation_main_test.dart:
```diff
@@ -0,0 +1,49 @@
+  test('cutout should be on top in portrait mode', () async {
+    await nativeDriver.rotateResetDefault();
+    await flutterDriver.waitFor(find.byValueKey('CutoutTop'),
+        timeout: const Duration(seconds: 5));
```
How was a timeout of `seconds: 5` decided and why was it necessary?

yaakovschectman replied:
The exact value is somewhat arbitrary, but the timeout is there because that is essentially the failure condition of the test. If the orientation detection fails and the corresponding widget key is never present, this is when the test should fail.

matanlurey commented on dev/integration_tests/native_driver_test/lib/display_cutout_rotation_main.dart:
```diff
@@ -0,0 +1,59 @@
+    } else if (displayFeatures.length > 1) {
+      text = 'CutoutMany';
+    } else {
+      final Rect cutout = displayFeatures[0].bounds;
```
`displayFeatuers.single` would describe the intent better.

matanlurey reviewed (changes requested):
I've left some comments below _if_ we go with this approach. However, I'm not sure it's the right approach - `native_driver_test` was intended to test things that are otherwise not practical or possible to test with more standard testing, but as far as I can tell we could have used `integration_test` and written a standard integration test to test this:

<https://docs.flutter.dev/testing/integration-tests>

For example, then we wouldn't need the awkward `waitFor`, as it would be running in-app process. Even the rotation can be done using in-app process functionality, i.e. using `SystemChrome`: https://github.com/flutter/flutter/blob/d6918a48d3bb8c247efabea04b3361e1c4a40e02/dev/tools/native_driver/lib/extension.dart#L76-L78.

So, unless there is a strong reason to do so, I'd rather see this as a (new) test in `dev/integration_tests` specifically made for this piece of functionality. The comments I've left below are for if there is a strong reason to do so.

yaakovschectman commented:
@matanlurey I started looking at the native_driver tests and their screen rotation methods after [your comment](https://github.com/flutter/engine/pull/55992#discussion_r1817379177) on the engine PR about rotation tests. But it will still be possible to relocate it.

reidbaker committed:
Inital display cutout integration tests app host folder

reidbaker committed:
Custom driver, still need to modify app code to be less complex

reidbaker commented:
I am taking over this pr. Newly added test can be run with flutter drive.

reidbaker committed:
Merge branch 'master' into android_it_rot

reidbaker commented:
Replaced by https://github.com/flutter/flutter/pull/160255
